File: src/main/scala/com/weave/core/GraphEvent.scala
package com.weave.core

sealed trait GraphEvent {
  val name: String
}

object GraphEvent {

  case class NodeStarted(
      name: String
  ) extends GraphEvent

  case class NodeCompleted(
      name: String
  ) extends GraphEvent

  case class WorkflowCompleted(name: String) extends GraphEvent

  case class NodeFailed(
      name: String,
      cause: Throwable
  ) extends GraphEvent

  case class WorkflowFailed(
      name: String,
      cause: Throwable
  ) extends GraphEvent

  case class CheckpointCreated[S](
      name: String,
      state: S
  ) extends GraphEvent
}

File: src/main/kotlin/com/weave/core/GraphEvent.kt
package com.weave.core

public sealed interface GraphEvent {
    public val name: String

    public data class NodeStarted(override val name: String) : GraphEvent

    public data class NodeCompleted(override val name: String) : GraphEvent

    public data class NodeFailed(
        override val name: String,
        public val cause: Throwable,
    ) : GraphEvent

    public data class CheckpointCreated<S>(
        override val name: String,
        public val state: S,
    ) : GraphEvent

    public data class WorkflowCompleted(override val name: String) : GraphEvent

    public data class WorkflowFailed(
        override val name: String,
        public val cause: Throwable,
    ) : GraphEvent
}
